README.md

[![NPM Version](https://img.shields.io/npm/v/reactivesocket.svg)](https://npmjs.org/package/reactivesocket)
[![Build Status](https://travis-ci.org/ReactiveSocket/reactivesocket-js.svg?branch=master)](https://travis-ci.org/ReactiveSocket/reactivesocket-js)
[![Coverage Status](https://coveralls.io/repos/github/ReactiveSocket/reactivesocket-js/badge.svg?branch=master)](https://coveralls.io/github/ReactiveSocket/reactivesocket-js?branch=master)

# reactivesocket-js
ReactiveSocket Protocol for Client/Server for JS. Also comes with a [CLI](#CLI).
```bash
npm install -g reactivesocket
```
This library only supports the `request/response`, `setup` and `error`
interactions. More interactions are coming soon.

## Streams
The transport for this library is built entirely on top of the Node.js
[Stream](https://nodejs.org/api/stream.html) API.  As a result, it is agnostic
to the underlying transport mechanism. As long as you pass in a transport
stream that is a Node.js
[Duplex](https://nodejs.org/api/stream.html#stream_class_stream_duplex) stream,
this library will work.

Using streams means that this library natively supports backpressure regardless
of the transport implementation.

We currently target TCP via the [net](https://nodejs.org/api/net.html) module,
and WebSockets via the [yws-stream](https://github.com/yunong/ws-stream)
module. You are of course, free to inject other transports.

## Connection Quick Start
### TCP
```javascript
var net = require('net');

var bunyan = require('bunyan');
var reactiveSocket = require('reactivesocket');


// Create any transport stream that's a Node.js Duplex Stream.
var transportStream = net.connect(1337, 'localhost', function (err) {
    var rsConnection = reactiveSocket.createConnection({
        log: bunyan.createLogger({name: 'rsConnection'}),
        transport: {
            stream: transportStream,
            framed: true // TCP requires explicit framing
        },
        type: 'client',
        metadataEncoding: 'utf8',
        dataEncoding: 'utf8'
    });

    rsConnection.on('ready', function () {
        // returns a reactive socket stream
        var stream = rsConnection.request({
            metadata: 'You reached for the secret too soon, you cried for the moon',
            data: 'Shine on you crazy diamond.'
        });

        stream.on('response', function (res) {
            console.log('got response', res.getResponse());
        });

        stream.on('application-error', function (err) {
            console.error('got error', err);
        });
    });
});
```
### WebSocket
```javascript
var bunyan = require('bunyan');
var reactiveSocket = require('reactivesocket');

var Ws = require('ws');
var WSStream = require('yws-stream');


var websocket = new Ws('ws://localhost:1337');

// Create any transport stream that's a Node.js Duplex Stream
var transportStream = new WSStream({
    log: bunyan.createLogger({name: 'ws-stream'}),
    ws: websocket
});

// Wait for Websocket to establish connection, before we create an RS Connection
websocket.on('open', function() {

    var rsConnection = reactiveSocket.createConnection({
        log: bunyan.createLogger({name: 'rsConnection'}),
        transport: {
            stream: transportStream
        },
        type: 'client',
        metadataEncoding: 'utf8',
        dataEncoding: 'utf8'
    });

    rsConnection.on('ready', function () {
        // returns a reactive socket stream
        var stream = rsConnection.request({
            metadata: 'You reached for the secret too soon, you cried for the moon',
            data: 'Shine on you crazy diamond.'
        });

        stream.on('response', function (res) {
            console.log('got response', res.getResponse());
        });

        stream.on('application-error', function (err) {
            console.error('got error', err);
        });
    });
});
```

## CLI
This library comes with a CLI. You can use it by installing this module.
```bash
$ npm install -g reactivesocket
```

### RS Client
There are two versions of the client CLI. The simple CLI makes one request to a
server.
```bash
$ rs -o req tcp://localhost:1337 'if you didnt care what happened to me, And I didnt care for you'
```

There is also a benchmarking CLI in the vein of [Apache
Bench](https://httpd.apache.org/docs/2.4/programs/ab.html)
```bash
$ rb -c 10 -n 10000000 -s 1000 tcp://localhost:1337
{ 'elapsed time (s)': 10.529176232,
  'total reqs': 137133,
  RPS: 13024.095805636622,
  'median (ms)': 0.649035,
  'mean (ms)': 0.75758988656268,
  '0.1% (ms)': 0.457949,
  '1% (ms)': 0.498248,
  '5% (ms)': 0.544133,
  '10% (ms)': 0.565295,
  '20% (ms)': 0.596515,
  '30% (ms)': 0.616699,
  '40% (ms)': 0.633112,
  '50% (ms)': 0.649035,
  '60% (ms)': 0.671943,
  '70% (ms)': 0.708819,
  '80% (ms)': 0.772095,
  '90% (ms)': 0.905283,
  '99% (ms)': 4.441137,
  '99.9% (ms)': 6.004325,
  '99.99% (ms)': 32.613085,
  '99.999% (ms)': 101.189893 }
```

### Echo Servers
Simple echo servers are also available for both TCP and Websocket.

#### TCP
```bash
$ HOST=localhost PORT=1337 rs-tcp-server
```
#### WebSocket
```bash
$ HOST=localhost PORT=1337 rs-ws-server
```

## Contributions
Contributions welcome, please ensure `make check` runs clean.

## License
MIT

Copyright 2016 Yunong J Xiao


CHANGELOG.md
# 1.2.0
- Added CLI support
- Performance fixes

<a name="0.2.0"></a>
# [0.2.0](https://github.com/ReactiveSocket/reactivesocket-js/compare/v0.1.0...v0.2.0) (2016-03-08)


### Bug Fixes

* **memory:** fill(0) was forgotten in expected data. ([f975b7a](https://github.com/ReactiveSocket/reactivesocket-js/commit/f975b7a))
* **ws:** There were some missing exports. ([e5fd535](https://github.com/ReactiveSocket/reactivesocket-js/commit/e5fd535))

### Features

* **ws-transport:** The websocket transport ([48bb5c8](https://github.com/ReactiveSocket/reactivesocket-js/commit/48bb5c8))



<a name="0.0.2"></a>
## [0.0.2](https://github.com/ReactiveSocket/reactivesocket-js/compare/v0.0.1...v0.0.2) (2016-02-26)


### Features

* **factory:** Added the factory functions for TCP connections ([e3a1763](https://github.com/ReactiveSocket/reactivesocket-js/commit/e3a1763))



<a name="0.0.1"></a>
## 0.0.1 (2016-02-26)


### Features

* **client:** The ReactiveClient: happy case only ([2f4f9b8](https://github.com/ReactiveSocket/reactivesocket-js/commit/2f4f9b8))
* **errorFrame:** Tested the error frame ([762b375](https://github.com/ReactiveSocket/reactivesocket-js/commit/762b375))
* **frames:** Frame header and setup frame. ([3b78f7c](https://github.com/ReactiveSocket/reactivesocket-js/commit/3b78f7c))
* **metaflag:** There was an issue with metadata. ([7f29cce](https://github.com/ReactiveSocket/reactivesocket-js/commit/7f29cce))
* **request-response:** The Request/Response Frame ([f5fb2fb](https://github.com/ReactiveSocket/reactivesocket-js/commit/f5fb2fb))
* **respond:** The server now has a way to respond to ReqRes ([520ebf9](https://github.com/ReactiveSocket/reactivesocket-js/commit/520ebf9))
* **response:** The response frame and tests ([52c5312](https://github.com/ReactiveSocket/reactivesocket-js/commit/52c5312))
* **server:** The happy-case server ([22a377d](https://github.com/ReactiveSocket/reactivesocket-js/commit/22a377d))
* **tcp:** The TCP transport layer for client/server. ([430fc00](https://github.com/ReactiveSocket/reactivesocket-js/commit/430fc00))
* **travis:** Travis, Coveralls, and Bithound integration ([680c7f0](https://github.com/ReactiveSocket/reactivesocket-js/commit/680c7f0))





perf/rsBench.js
'use strict';

var fs = require('fs');
var net = require('net');
var url = require('url');

var _  = require('lodash');
var async = require('async');
var dashdash = require('dashdash');
var ss = require('simple-statistics');
var vasync = require('vasync');

var Ws = require('ws');
var WSStream = require('yws-stream');

var reactiveSocket = require('../lib');

var TCP_REGEX = /tcp:\/\/.+:[0-9]+\/*.*/;
var WS_REGEX = /wss?:\/\/.+:[0-9]+\/*.*/;

var OPTIONS = [{
    names: ['help', 'h'],
    type: 'bool',
    help: 'Print this help and exit.'
},{
    names: ['number', 'n'],
    type: 'number',
    help: 'Number of requests to perform'
}, {
    names: ['concurrency', 'c'],
    type: 'number',
    help: 'Number of multiple requests to make at a time'
}, {
    names: ['timelimit', 't'],
    type: 'number',
    help: 'Seconds to max. to spend on benchmarking'
}, {
    names: ['size', 's'],
    type: 'number',
    help: 'Size of payload in bytes, defaults to 8'
}, {
    names: ['metadata', 'm'],
    type: 'string',
    help: 'Metadata to include, defaults to null'
}];

var PARSER = dashdash.createParser({options: OPTIONS});
var HELP = PARSER.help({includeEnv: true}).trimRight();
var OPTS;

try {
    OPTS = PARSER.parse(process.argv);
} catch (e) {
    console.error('foo: error: %s', e.message);
    process.exit(1);
}

if (OPTS.help) {
    HELP = PARSER.help({includeEnv: true}).trimRight();
    console.log('usage: rsBench [OPTIONS] tcp|ws://localhost:1337\n'
                + 'options:\n'
                + HELP);
    process.exit(0);
}

var RAW_URL = OPTS._args[0];

var ITERATIONS = OPTS.number || 1000;
var CONCURRENCY = OPTS.concurrency || 10;
var SIZE = OPTS.size || 8;
var ENDPOINT = url.parse(RAW_URL);

// we need to send a large enough frame to ensure we exceed the default TCP
// loopback MTU of 16-64 kbytes. This is to test that framing actually works.
// Hence we read in some select works of the Bard.
var DATA = fs.readFileSync('./test/etc/hamlet.txt', 'utf8');

for (var i = 0; i < SIZE / DATA.length ; i++) {
    DATA += DATA;
}

DATA = DATA.substr(0, SIZE);

var RS_CLIENT_CON;
var CLIENT_STREAM;

var REQ = {
    metadata: OPTS.metadata,
    data: DATA
};

var TIMERS = new Array(ITERATIONS);

var COUNT = 0;

var START_TIME;
var HAS_PRINTED = false;

vasync.pipeline({funcs: [
    function setupTransportStream(ctx, cb) {
        if (TCP_REGEX.test(RAW_URL)) {
            CLIENT_STREAM = net.connect(ENDPOINT.port, ENDPOINT.hostname,
                                        function (e) {
                return cb(e);
            });
        } else if (WS_REGEX.test(RAW_URL)) {
            var ws = new Ws(RAW_URL);
            CLIENT_STREAM = new WSStream({
                ws: ws
            });

            ws.on('open', function () {
                return cb();
            });

        } else {
            HELP = PARSER.help({includeEnv: true}).trimRight();
            console.log('usage: rsBench [OPTIONS] tcp|ws://localhost:1337\n'
                        + 'options:\n'
                        + HELP);
            process.exit(1);
        }
    },
    function setupConnection(ctx, cb) {
        RS_CLIENT_CON = reactiveSocket.createConnection({
            transport: {
                stream: CLIENT_STREAM,
                framed: true
            },
            type: 'client',
            metadataEncoding: 'utf-8',
            dataEncoding: 'utf-8'
        });

        RS_CLIENT_CON.on('ready', function () {
            START_TIME = process.hrtime();
            async.eachLimit(TIMERS, CONCURRENCY, function (timer, _cb) {
                var start = process.hrtime();
                var stream = RS_CLIENT_CON.request(REQ);

                stream.on('response', function (res) {
                    var elapsed = process.hrtime(start);
                    var elapsedNs = elapsed[0] * 1e9 + elapsed[1];
                    TIMERS[COUNT] = elapsedNs;
                    COUNT++;
                    _cb();

                    if (COUNT === ITERATIONS) {
                        return cb();
                    }
                });
            });
        });
    }
], arg: {}}, function (err, cb) {
    if (err) {
        throw err;
    }

    process.exit();
});

function printMetrics() {
    var timers = _.compact(TIMERS);
    var elapsed = process.hrtime(START_TIME);
    var elapsedNs = elapsed[0] * 1e9 + elapsed[1];

    if (HAS_PRINTED) {
        return;
    }
    HAS_PRINTED = true;
    console.log('elapsed time', elapsedNs / 1e9);
    console.log('total reqs', timers.length);
    console.log('rps', timers.length / (elapsedNs / 1e9));
    console.log('median', ss.median(timers) / 1e9);
    console.log('mean', ss.mean(timers) / 1e9);
    console.log('0.1%', ss.quantile(timers, 0.001) / 1e9);
    console.log('1%', ss.quantile(timers, 0.01) / 1e9);
    console.log('5%', ss.quantile(timers, 0.05) / 1e9);
    console.log('10%', ss.quantile(timers, 0.1) / 1e9);
    console.log('20%', ss.quantile(timers, 0.2) / 1e9);
    console.log('30%', ss.quantile(timers, 0.3) / 1e9);
    console.log('40%', ss.quantile(timers, 0.4) / 1e9);
    console.log('50%', ss.quantile(timers, 0.5) / 1e9);
    console.log('60%', ss.quantile(timers, 0.6) / 1e9);
    console.log('70%', ss.quantile(timers, 0.7) / 1e9);
    console.log('80%', ss.quantile(timers, 0.8) / 1e9);
    console.log('90%', ss.quantile(timers, 0.9) / 1e9);
    console.log('99%', ss.quantile(timers, 0.99) / 1e9);
}

process.on('exit', function () {
    printMetrics();
});

process.on('SIGINT', function () {
    printMetrics();
    process.exit();
});
